File: midonet-cluster/src/main/java/org/midonet/client/resource/Bridge.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.client.resource;

import java.net.URI;
import java.util.ArrayList;
import java.util.UUID;

import javax.ws.rs.core.UriBuilder;

import org.midonet.client.WebResource;
import org.midonet.client.dto.DtoBridge;
import org.midonet.client.dto.DtoBridgePort;
import org.midonet.client.dto.DtoDhcpSubnet;
import org.midonet.client.dto.DtoDhcpSubnet6;
import org.midonet.client.dto.DtoMacPort;
import org.midonet.client.dto.DtoPort;
import org.midonet.client.dto.DtoRouterPort;
import org.midonet.cluster.services.rest_api.MidonetMediaTypes;

public class Bridge extends ResourceBase<Bridge, DtoBridge> {

    public Bridge(WebResource resource, URI uriForCreation, DtoBridge b) {
        super(resource, uriForCreation, b,
              MidonetMediaTypes.APPLICATION_BRIDGE_JSON_V4());
    }

    @Override
    public URI getUri() {
        return principalDto.getUri();
    }

    public UUID getId() {
        return principalDto.getId();
    }

    public UUID getInboundFilterId() {
        return principalDto.getInboundFilterId();
    }

    public String getName() {
        return principalDto.getName();
    }

    public boolean isAdminStateUp() {
        return principalDto.isAdminStateUp();
    }

    public UUID getOutboundFilterId() {
        return principalDto.getOutboundFilterId();
    }

    public String getTenantId() {
        return principalDto.getTenantId();
    }

    public Bridge name(String name) {
        principalDto.setName(name);
        return this;
    }

    public Bridge adminStateUp(boolean adminStateUp) {
        principalDto.setAdminStateUp(adminStateUp);
        return this;
    }

    public Bridge tenantId(String tenantId) {
        principalDto.setTenantId(tenantId);
        return this;
    }

    public Bridge inboundFilterId(UUID id) {
        principalDto.setInboundFilterId(id);
        return this;
    }

    public Bridge outboundFilterId(UUID id) {
        principalDto.setOutboundFilterId(id);
        return this;
    }

    public ResourceCollection<BridgePort> getPorts() {
        return getChildResources(
                principalDto.getPorts(),
                null,
                MidonetMediaTypes.APPLICATION_PORT_V3_COLLECTION_JSON(),
                BridgePort.class, DtoBridgePort.class);
    }

    public ResourceCollection<Port<?,?>> getPeerPorts() {
        ResourceCollection<Port<?,?>> peerPorts =
                new ResourceCollection<>(new ArrayList<Port<?,?>>());

        DtoPort[] dtoPeerPorts = resource.get(
                principalDto.getPeerPorts(),
                null,
                DtoPort[].class,
                MidonetMediaTypes.APPLICATION_PORT_V3_COLLECTION_JSON());

        for (DtoPort pp : dtoPeerPorts) {
            if (pp instanceof DtoRouterPort) {
                RouterPort rp = new RouterPort(
                        resource,
                        principalDto.getPorts(),
                        (DtoRouterPort) pp);
                peerPorts.add(rp);

            } else if (pp instanceof DtoBridgePort) {
                // Interior bridge ports can be linked if one of the 2 has a
                // vlan id assigned, this should be validated upon creation
                BridgePort bp = new BridgePort(
                        resource,
                        principalDto.getPorts(),
                        (DtoBridgePort) pp);
                peerPorts.add(bp);
            }
        }
        return peerPorts;
    }

    public BridgePort addPort() {
        return new BridgePort(
                resource,
                principalDto.getPorts(),
                new DtoBridgePort());
    }

    public DhcpSubnet addDhcpSubnet() {
        return new DhcpSubnet(resource, principalDto.getDhcpSubnets(),
                          new DtoDhcpSubnet());
    }

    public ResourceCollection<DhcpSubnet> getDhcpSubnets() {
        return getChildResources(
            principalDto.getDhcpSubnets(),
            null,
            MidonetMediaTypes.APPLICATION_DHCP_SUBNET_COLLECTION_JSON_V2(),
            DhcpSubnet.class, DtoDhcpSubnet.class);
    }


    public DhcpSubnet6 addDhcpSubnet6() {
        return new DhcpSubnet6(resource, principalDto.getDhcpSubnet6s(),
                          new DtoDhcpSubnet6());
    }

    public ResourceCollection<DhcpSubnet6> getDhcpSubnet6s() {
        return getChildResources(
            principalDto.getDhcpSubnet6s(),
            null,
            MidonetMediaTypes.APPLICATION_DHCPV6_SUBNET_COLLECTION_JSON(),
            DhcpSubnet6.class, DtoDhcpSubnet6.class);
    }

    private URI getMacTableUri(Short vlanId) {
        return (vlanId == null) ?
                principalDto.getMacTable() :
                createUriFromTemplate(principalDto.getVlanMacTableTemplate(),
                        "{vlanId}", vlanId);
    }

    public ResourceCollection<MacPort> getMacTable(Short vlanId) {
        return getChildResources(getMacTableUri(vlanId), null,
                MidonetMediaTypes.APPLICATION_MAC_PORT_COLLECTION_JSON_V2(),
                MacPort.class, DtoMacPort.class);
    }

    public MacPort addMacPort(Short vlanId, String macAddr, UUID portId) {
        DtoMacPort mp = new DtoMacPort(macAddr, portId);
        return new MacPort(resource, getMacTableUri(vlanId), mp);
    }

    public MacPort getMacPort(Short vlanId, String macAddr, UUID portId) {
        String uriMacAddr = macAddr.replace(':', '-');
        URI uri = (vlanId == null) ?
                UriBuilder.fromPath(principalDto.getMacPortTemplate())
                        .build(uriMacAddr, portId) :
                UriBuilder.fromPath(principalDto.getVlanMacPortTemplate())
                        .build(vlanId, uriMacAddr, portId);
        DtoMacPort mp = resource.get(uri, null, DtoMacPort.class,
                            MidonetMediaTypes.APPLICATION_MAC_PORT_JSON_V2());
        return new MacPort(resource, mp.getUri(), mp);
    }

    @Override
    public String toString() {
        return String.format("Bridge{id=%s, name=%s}", principalDto.getId(),
                             principalDto.getName());
    }
}

File: python-midonetclient/src/midonetclient/bridge.py
# vim: tabstop=4 shiftwidth=4 softtabstop=4

# Copyright 2013 Midokura PTE LTD.
# All Rights Reserved
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.


from midonetclient import admin_state_up_mixin
from midonetclient import dhcp_subnet
from midonetclient import port
from midonetclient import port_type
from midonetclient import resource_base
from midonetclient import vendor_media_type


class Bridge(resource_base.ResourceBase,
             admin_state_up_mixin.AdminStateUpMixin):

    media_type = vendor_media_type.APPLICATION_BRIDGE_JSON

    def __init__(self, uri, dto, auth):
        super(Bridge, self).__init__(uri, dto, auth)

    def get_name(self):
        return self.dto['name']

    def get_id(self):
        return self.dto['id']

    def get_tenant_id(self):
        return self.dto['tenantId']

    def get_inbound_filter_id(self):
        return self.dto['inboundFilterId']

    def get_outbound_filter_id(self):
        return self.dto['outboundFilterId']

    def name(self, name):
        self.dto['name'] = name
        return self

    def tenant_id(self, tenant_id):
        self.dto['tenantId'] = tenant_id
        return self

    def inbound_filter_id(self, id_):
        self.dto['inboundFilterId'] = id_
        return self

    def outbound_filter_id(self, id_):
        self.dto['outboundFilterId'] = id_
        return self

    def get_inbound_mirrors(self, query=None):
        return self.dto['inboundMirrorIds']

    def inbound_mirrors(self, inMirrors):
        self.dto['inboundMirrorIds'] = inMirrors
        return self

    def get_outbound_mirrors(self, query=None):
        return self.dto['outboundMirrorIds']

    def outbound_mirrors(self, outMirrors):
        self.dto['outboundMirrorIds'] = outMirrors
        return self

    def get_ports(self, query=None):
        if query is None:
            query = {}
        headers = {'Accept':
                   vendor_media_type.APPLICATION_PORT_COLLECTION_JSON}
        return self.get_children(self.dto['ports'], query, headers, port.Port)

    def get_peer_ports(self, query=None):
        if query is None:
            query = {}
        headers = {'Accept':
                   vendor_media_type.APPLICATION_PORT_COLLECTION_JSON}
        _, peer_ports = self.auth.do_request(self.dto['peerPorts'], 'GET',
                                             headers=headers, query=query)
        res = []
        for pp in peer_ports:
            res.append(port.Port(self.dto['ports'], pp, self.auth))
        return res

    def get_dhcp_subnets(self):
        query = {}
        headers = {'Accept':
                   vendor_media_type.APPLICATION_DHCP_SUBNET_COLLECTION_JSON}
        return self.get_children(self.dto['dhcpSubnets'], query, headers,
                                 dhcp_subnet.DhcpSubnet)

    def get_dhcp_subnet(self, subnet_str):
        """
        given the subnet_str (x.x.x.x_y, address_length), returns DhcpSubnet
        resource that has subnet prefix and length found in the subnet_str.
        """
        prefix, length = subnet_str.split('_')
        for ds in self.get_dhcp_subnets():
            if ds.get_subnet_prefix() == prefix and \
                    ds.get_subnet_length() == int(length):
                return ds
        return None

    def add_port(self):
        return port.Port(self.dto['ports'],
                    {'type': port_type.BRIDGE}, self.auth)

    def add_dhcp_subnet(self):
        return dhcp_subnet.DhcpSubnet(self.dto['dhcpSubnets'], {}, self.auth)
